homeassistant/components/mqtt/__init__.py

```diff
@@ -370,18 +389,30 @@ def async_publish_service(call):
         msg_topic = call.data[ATTR_TOPIC]
         payload = call.data.get(ATTR_PAYLOAD)
         payload_template = call.data.get(ATTR_PAYLOAD_TEMPLATE)
+        payload_file_path = call.data.get(ATTR_PAYLOAD_FILE_PATH)
         qos = call.data[ATTR_QOS]
         retain = call.data[ATTR_RETAIN]
-        try:
-            if payload_template is not None:
+        if payload_template is not None:
+            try:
                 payload = \
                     template.Template(payload_template, hass).async_render()
-        except template.jinja2.TemplateError as exc:
-            _LOGGER.error(
-                "Unable to publish to '%s': rendering payload template of "
-                "'%s' failed because %s",
-                msg_topic, payload_template, exc)
-            return
+            except template.jinja2.TemplateError as exc:
+                _LOGGER.error(
+                    "Unable to publish to '%s': rendering payload template of "
+                    "'%s' failed because %s",
+                    msg_topic, payload_template, exc)
+                return
+        elif payload_file_path is not None:
```

Comment:
So please advice how a component could send a binary payload if the underlying MQTT implementation does not allow for it. The way I see it, HA MQTT core implementation  does not conform to MQTT specifications [here](http://public.dhe.ibm.com/software/dw/webservices/ws-mqtt/mqtt-v3r1.html) and [here](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html):  

> A messaging transport that is agnostic to the content of the payload.

base64 encoding is a workaround and, if that is the answer, this whole PR is quite unnecessary, because it doesn't make HA compliant to the standard, which was my original goal.
Thanks